url71665d7d68_branscha_lib_scripty_tgz-pJ8-com_sdicons_scripty_ScriptyAutoReplJ8/src/com/sdicons/scripty/parser/Eval2.java

```diff
@@ -0,0 +1,1640 @@
+                    aStack.push(lLstExpr.get(2), lCtx);
+                    return;
+                }
+                else
+                {
+                    // While expression without body.
+                    // Move to slot 0 to receive the result of boolean evaluation.
+                    aFrame.setDataPtr(0);
+                    // There is no body, we re-evaluate the boolean expression.
+                    aStack.push(lLstExpr.get(1), lCtx);
+                    return;
+                }
+            }
+            else
+            {
+                // Done.
+                aFrame.setResult(lData[1]);
+                return;
+            }
+        default:
+            // Re-evaluate the boolean expression.
+            aFrame.setDataPtr(0);
+            aStack.push(lLstExpr.get(1), lCtx);
+        }
+    }
+}
+
+class AndHandler
+implements Eval2.IFrameHandler
+{
+    public void init(Eval2.EvalStack aStack, Eval2.StackFrame aFrame)
+    throws CommandException
+    {
+        final List lLstExpr = (List) aFrame.getExpr();
+        final int lLstSize = lLstExpr.size();
+
+        if(lLstSize <= 1) throw new CommandException("The 'and' form should have the format (and <bool-expr>+).", aStack);
+
+        //Allocate frame space. We allocate enough space to evaluate all arguments.
+        aFrame.allocateData(lLstSize - 1);
+    }
+
+    public void handleFrame(IEval aEval, Eval2.EvalStack aStack, Eval2.StackFrame aFrame)
+    throws CommandException
+    {
+        // It is a special form because it shortcuts evaluation if it encounters a
+        // false value.
+
+        final List lLstExpr = (List) aFrame.getExpr();
+        final int lLstSize = lLstExpr.size();
+        final IContext lCtx = aFrame.getCtx();
+        final Object[] lData = aFrame.getData();
+        final int lDataPtr = aFrame.getDataptr();
+
+        switch(lDataPtr)
+        {
+        case 0:
+            // Push the first expression on the stack for evaluation.
+            aStack.push(lLstExpr.get(1), lCtx);
+            return;
+        default:
+            // Check the result of the last parameter.
+            if(!Eval2.boolEval(lData[lDataPtr - 1]))
+            {
+                // We shortcut the evaluation if we find a false value.
+                // We can stop on the first false value we encounter.
+                aFrame.setResult(Boolean.FALSE);
+            }
+            else
+            {
+                if(lDataPtr < (lLstSize - 1))
+                {
+                    // There are still some and-expressions to evaluate, so we
+                    // evaluate the next one.
+                    aStack.push(lLstExpr.get(lDataPtr + 1), lCtx);
+                }
+                else
+                {
+                    // We evaluated all expressions, and the all evaluated
+                    // to true.
+                    aFrame.setResult(Boolean.TRUE);
+                }
+            }
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1031:1066, 1091:1126.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.